### DISCUSSION TITLE
Custom Migrations and Directus schema

### DISCUSSION THREAD

**Discussion by abdonrd** [directus/directus#11798] [Archive]
Following the official Directus documentation about [Migrations and Directus schema](https://docs.directus.io/extensions/migrations/#migrations-and-directus-schema).

And I have my migration file at: `extensions/migrations/20220222A-bootstrap.js`

To setup a new environment, I run a npm script that run:

```
"start": "npx directus database install && npx directus schema apply snapshot.yaml && npx directus database migrate:latest"
```

1. `npx directus database install`
  Directus install all the initial schema

2. `npx directus schema apply snapshot.yaml`
  Directus should apply my snapshot

3. `npx directus database migrate:latest`
  Directus should apply all the migrations

But I have an error with the step 2.
Because my `snapshot.yaml` has references to `accountability`.

> 🚨 error select "collection", "singleton", "note", "sort_field", "accountability" from "directus_collections" - column "accountability" does not exist

And with just the step 1, the database don't have all the Directus migrations from:
https://github.com/directus/directus/tree/main/api/src/database/migrations

For example the related with the accountability:
https://github.com/directus/directus/blob/main/api/src/database/migrations/20210416A-add-collections-accountability.ts

So... There is any option to just apply the Directus migrations?

**Comment by abdonrd** [comment]
Oh, and now I have another error:

When the database is already installed, the `npx directus database install` command throws an error.

So maybe I'm not following a step correctly.

**Comment by rijkvanzanten** [comment]
I think it's due to the order of operations here. In step 2, you're applying the database snapshot that was created against the latest version of Directus on top of a database that's technically at v9.0.0-alpha.1. This can cause all sorts of issues, as the Directus installation that applies the snapshot expects the database to be up to date with the current version of Directus. You'll have to run the migrations first, and then apply the snapshot.

This is where it gets interesting: you effectively want to install and migrate the database fresh to the latest versions, apply the snapshot, and _then_ apply your custom migrations at the end. This use case and discussion has popped up before, so I'm very curious to hear your thoughts on that flow and how to potentially solve for that. 

One potential solution is to add a "--skip-custom" flag to the `database migrate:latest` and/or bootstrap commands, so you can do something like `npx directus bootstrap && npx directus schema apply && npx directus database migrate:latest`.

cc @keesvanbemmel

**Reply by abdonrd** [reply]
Maybe something like this?

1. `npx directus bootstrap` (I don't know with or without `--skip-custom`)
  Directus install all the initial schema, **and run only the Directus migrations**.

2. `npx directus schema apply snapshot.yaml`
  Directus should apply the snapshot

3. `npx directus database migrate:latest`
  Directus should apply all the migrations

**Reply by rijkvanzanten** [reply]
Yeah! We might even want to make a custom "migrate" command that only runs custom migrations, that way, we don't need any flags or complexity. It could be something like `npx directus database migrate:custom:latest`

**Reply by abdonrd** [reply]
I like it!

**Reply by abdonrd** [reply]
But I think it would be nice if `npx directus bootstrap` internally ran the Directus core migrations by default.

**Reply by rijkvanzanten** [reply]
Yeah I agree 👍🏻  It sounds like the main pain point that comes up in these discussions is the fact that that initial bootstrap/migrate command runs the custom migrations as well, so lets adjust that 🙂

**Reply by mlbiche** [reply]
Hey there 👋  I am also interested in this point and my search brought me here. Has it been implemented yet ? Is there an opened issue to track this feature ? Cheers !

**Reply by rijkvanzanten** [reply]
> Hey there 👋 I am also interested in this point and my search brought me here. Has it been implemented yet ? Is there an opened issue to track this feature ? Cheers !

This is the discussion to improve the setup, so this is the place to track 🙂

**Reply by mlbiche** [reply]
I've just noticed this [part of the documentation](https://docs.directus.io/extensions/migrations/#migrations-and-directus-schema). For me, it is already solving the problem in fact ? 🤔

**Reply by scottrepreneur** [reply]
Was there a workaround for this? I'm running into trouble on the `schema apply` step also. Also in docker. 

Attempting to replace `npx directus bootstrap` so [custom migrations](https://docs.directus.io/extensions/migrations.html#migrations-and-directus-schema) are run in the correct order. Is there something else that needs to be run before the snapshot is applied?

```docker
# replacing npx directus bootstrap
CMD npx directus database install \
    && npx directus schema apply --yes ./snapshot.yaml \
    && npx database migrate:latest \
    && npx directus start
```

**Reply by azrikahar** [reply]
I think that section of the docs may need to add some caveat, or potentially remove the snapshot step since it may introduce confusions as seen in this thread 🤔

> Was there a workaround for this? 

The closest workaround for now might be to copy your custom migrations to `extensions/migrations` _after_ doing bootstrap, so the steps may look like:

1. Ensure your custom migrations are NOT included in `extensions/migrations` beforehand
2. `npx directus bootstrap`
3. `npx directus schema apply --yes ./snapshot.yaml`
4. `mv` or `cp` your custom migration files to `extensions/migrations`
5. `npx directus database migrate:latest`

---

The addition of a potential `--skip-custom` flag as discussed in the parent comment above essentially helps to remove the need to do this "move/copy your custom migration files after boostrap" extra step.

**Reply by mlbiche** [reply]
Is there any move planned for this flag?

**Reply by stafyniaksacha** [reply]
> The closest workaround for now might be to copy your custom migrations to extensions/migrations after doing bootstrap, so the steps may look like:


What about this being default bootstrap behavior when adding a `--snapshot <path>`  ?

**Reply by abdonrd** [reply]
@rijkvanzanten @azrikahar are you open for a `--skip-custom` PR?

To finally allow:

1. `npx directus bootstrap --skip-custom`
  Directus install all the initial schema, and run **only the Directus migrations**

2. `npx directus schema apply snapshots/latest.yaml`
  Directus apply the snapshot

3. `npx directus database migrate:latest`
  Directus apply all the migrations
  And the Directus migrations will be ignored because they have already been applied in step 1

4. `npx directus start`

Example with docker compose:

```
version: '3'
services:
  [...]
  directus:
    container_name: directus
    image: directus/directus:9.22.4
    [...]
    command:
      - /bin/sh
      - -c
      - |
        npx directus bootstrap --skip-custom &&
        npx directus schema apply snapshots/latest.yaml --yes &&
        npx directus database migrate:latest &&
        npx directus start
```

Thanks!

**Reply by badrange** [reply]
That would be helpful to us! The flag could also be `--only-directus-migrations` 

The suggestion about moving the custom migrations in place won't work for us because the container is mounted as read-only.

**Reply by paescuj** [reply]
Before proceeding to a PR, I'd like to go back to this original question:

> This is where it gets interesting: you effectively want to install and migrate the database fresh to the latest versions, apply the snapshot, and then apply your custom migrations at the end. This use case and discussion has popped up before, so I'm very curious to hear your thoughts on that flow and how to potentially solve for that.

AFAICS, this has not been answered yet.

The migrations are applied in the order of the dates. That is, if the custom migrations are named correctly, the problem shouldn't ever arise at all, right? See also https://docs.directus.io/extensions/migrations.html#file-name.

**Reply by badrange** [reply]
Custom migrations depending on tables that will be created when the schema is applied will fail.

**Reply by paescuj** [reply]
Okay, thanks! That sounds like a valid use case to me.

However, I'm afraid that this won't be properly solved by simply adding a `--skip-custom` option.
IIRC, Directus keeps track of the current state of the DB / migrations by storing the date of the last applied migration.
That is, when we have a flow like this
```bash
npx directus bootstrap --skip-custom
npx directus schema apply snapshots/latest.yaml
npx directus database migrate:latest
```
there's no guarantee that the custom migrations will actually be applied since newer "core" migrations might have been added in the meantime.

So the first thing we need to do here is to figure out how we can actually track the state for such custom migrations. 

**Edit**:
Maybe adding a separate state for custom migrations? We'd probably also have to distinguish between custom migrations that fit into the normal flow (depending on core migrations) and those that depend on a snapshot.

**Reply by abdonrd** [reply]
@paescuj as @badrange mention, when you have a custom migration depending on tables that will be created with the snapshot schema will fail.

Edit: We have written at the same time, now I will read your answer (https://github.com/directus/directus/discussions/11798#discussioncomment-4786348).

**Reply by abdonrd** [reply]
> there's no guarantee that the custom migrations will actually be applied since newer "core" migrations might have been added in the meantime.

Oh, good catch! I hadn't realized that.

> So the first thing we need to do here is to figure out how we can actually track the state for such custom migrations.
> 
> **Edit**: Maybe adding a separate state for custom migrations? We'd probably also have to distinguish between custom migrations that fit into the normal flow (depending on core migrations) and those that depend on a snapshot.

That reminds me of this comment from Rijk van Zanten at the beginning of this discussion: https://github.com/directus/directus/discussions/11798#discussioncomment-2267811

> Yeah! We might even want to make a custom "migrate" command that only runs custom migrations, that way, we don't need any flags or complexity. It could be something like `npx directus database migrate:custom:latest`

**Reply by abdonrd** [reply]
So maybe something like this?

1. `npx directus bootstrap --skip-custom`
  Directus install all the initial schema, and run **only the Directus migrations**

2. `npx directus schema apply snapshots/latest.yaml`
  Directus apply the snapshot

3. `npx directus database migrate:custom:latest`
  Directus apply all the **custom migrations**

4. `npx directus start`

**Reply by paescuj** [reply]
We still have to take into account that there are custom migrations which are integrated into the normal flow. Those that are dependent on the core migrations and must be applied in the correct order.
Perhaps we could distinguish them using a variable within the migration file or by the file name?
And somehow we need a way to track the state of those custom migrations that are applied separately.

**Reply by abdonrd** [reply]
> We still have to take into account that there are custom migrations which are integrated into the normal flow. Those that are dependent on the core migrations and must be applied in the correct order. Perhaps we could distinguish them using a variable within the migration file or by the file name? And somehow we need a way to track the state of those custom migrations that are applied separately.

Ok! I think it would work!

Something like:

flag = any way, like using a variable or by file name, to distinguish custom migrations that needs the schema

1. `npx directus bootstrap`
  Directus install all the initial schema, and run the migrations (except some under the flag)

2. `npx directus schema apply snapshots/latest.yaml`
  Directus apply the snapshot

3. `npx directus database migrate:flag:latest`
  Directus apply all the custom migrations under a flag

4. `npx directus start`

**Reply by MortensenC** [reply]
This is what I did to solve this:
I created a new folder migrationsWorkaround and moved all the migrations there.
<img width="252" alt="image" src="https://user-images.githubusercontent.com/435573/219469868-6b16a91c-6a9c-41ce-af3f-340a7b2fbd16.png">
Added these commands to package.json
`"schema:apply": "npm run directus -- schema apply --yes ./schema.yaml",`
`"move:migrations": "cp -r ./extensions/migrationsWorkaround/* ./extensions/migrations",`
`"clear:migrations": "find ./extensions/migrations -maxdepth 1 -type f -exec rm {} \\;",`

and on start command added this:
`npm run migrate:latest && npm run schema:apply && npm run move:migrations && npm run migrate:latest && npm run clear:migrations`

**Reply by abdonrd** [reply]
Sharing our current `Dockerfile`:

```dockerfile
FROM node:18-alpine AS builder

WORKDIR /builder

COPY package.json package-lock.json ./
COPY extensions/ extensions/
RUN npm ci && \
  npm run build

FROM directus/directus:9.23.1

# Switch to user 'node' and directory '/directus'
USER node
WORKDIR /directus

COPY snapshots/ ./snapshots/
COPY extensions/migrations-custom/ ./extensions/migrations-custom/
COPY --from=builder /builder/extensions/directus-extension-extension-name/ ./extensions/directus-extension-extension-name/
# COPY --from=builder /builder/extensions/directus-extension-extension-name-2/ ./extensions/directus-extension-extension-name-2/

CMD npx directus bootstrap && \
  npx directus schema apply snapshots/latest.yaml --yes && \
  mkdir -p extensions/migrations/ && \
  cp -r extensions/migrations-custom/* extensions/migrations/ && \
  npx directus database migrate:latest && \
  npx directus start
```

**Reply by noahjahn** [reply]
I feel bad for holding on to this implementation. I have been using some custom scripting around the database install/migrate commands. I added a few flags to handle the exact situation you described.

I created a PR for this! https://github.com/directus/directus/pull/17737

**Reply by hgezim** [reply]
@rijkvanzanten, the [docs](https://docs.directus.io/extensions/migrations.html#migrations-and-directus-schema) suggest running the migrations after applying snapshots:

```
npx directus database install
# notice that schema is applied before running migrations
npx directus schema apply ./path/to/﻿snapshot﻿.yaml
npx directus database migrate:latest
```

**Comment by seegeoff** [comment]
bump

**Comment by hgezim** [comment]
So, this is a midwit comment but why not just seed the database with a backed up snapshot instead of going through the gymnastics of migrations and schema changes?

```
FROM postgres:15.2

COPY ./seed.sql /docker-entrypoint-initdb.d/init.sql 

EXPOSE 5432
```

**Reply by rijkvanzanten** [reply]
> So, this is a midwit comment but why not just seed the database with a backed up snapshot instead of going through the gymnastics of migrations and schema changes?
> 
> ```
> FROM postgres:15.2
> 
> COPY ./seed.sql /docker-entrypoint-initdb.d/init.sql 
> 
> EXPOSE 5432
> ```

Seeds are a great starting point for a new database; migrations are often used to have a code-based way to move (an) existing database(s) to a new state 🙂

**Comment by jclaveau** [comment]
Maybe my similar experience could help people:
Only in CI context, starting my stack with `docker compose` using Directus image 10.3. When starting my stack like this
```yaml
    - name: Start the stack
      run: npm run ci

    - run: wait-port "$(docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' directus)":8055
      timeout-minutes: 1

    - name: Setting up the DB schema
      working-directory: ./apps/directus
      run: npm run directus:schema:apply -- --yes

```

I got intermittent database errors:
```
ERROR: select "id", "collection", "field", "special", "note", "validation" from "directus_fields" - column "validation" does not exist
```
and
```
ERROR: select "collection", "singleton", "note", "sort_field", "accountability" from "directus_collections" - column "accountability" does not exist
```

I fixed it by adding a simple delay:
```yaml
    - run: wait-port "$(docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' hiphiphip-directus)":8055
      timeout-minutes: 1

    - name: Setting up the DB schema
      working-directory: ./apps/directus
      # Sleep is required after the 8055 port is opened before applying schema to avoid
      # ERROR: select "collection", "singleton", "note", "sort_field", "accountability" from "directus_collections" - column "accountability" does not exist
      run: |
        sleep 5
        npm run directus:schema:apply -- --yes
```


+ Despite extending Directus docker image to settle the schema / seeds before `directus start`, is there a more reliable to check the directus instance readyness?
+ As these are false negative, is there an issue (something async which shouldn't be for example) in the starting pipe of Directus?
